segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,741 @@
+     */
+    public void concat(String segmentName, String sourceSegment, long offset) throws StreamSegmentException {
+        List<CuratorOp> curatorOps;
+        try {
+        curatorOps = this.getZKOperationsForConcat(segmentName, sourceSegment, offset);
+            List<CuratorTransactionResult> results = zkClient.transaction()
+                                                             .forOperations(curatorOps);
+            LogStorage logStorage = getOrRetrieveStorageLedger(segmentName);
+            logStorage.incrementUpdateVersion();
+            /* Fence all the ledgers and add one at the end. */
+            fenceLedgersAndCreateOneAtEnd(segmentName, logStorage);
+            /** Delete metadata and cache for the source once the operation is complete. */
+            this.zkClient.delete().deletingChildrenIfNeeded().forPath(getZkPath(sourceSegment));
+            ledgers.remove(sourceSegment);
+        } catch (Exception exc) {
+            /** In case of exception, drop the corrupt cache. */
+            ledgers.remove(segmentName);
+            throw translateZKException(segmentName, exc);
+        }
+    }
+
+    /**
+     * Deletes a segment along with all the data and metadata.
+     * @param segmentName name of the segment to be deleted.
+     * @return A CompletableFuture which completes once the delete is complete.
+     */
+    public void delete(String segmentName) throws StreamSegmentException {
+        LogStorage ledger = this.getOrRetrieveStorageLedger(segmentName);
+        if (ledger.getContainerEpoch() > this.containerEpoch.get()) {
+            throw new CompletionException(new StorageNotPrimaryException(segmentName));
+        }
+        try {
+            this.zkClient.delete()
+                         .deletingChildrenIfNeeded()
+                         .forPath(getZkPath(segmentName));
+        } catch (Exception e) {
+           log.warn("Exception while deleting a segment {}", segmentName, e);
+
+        }
+        ledgers.remove(segmentName);
+        deleteAllLedgers(ledger);
+    }
+
+    void deleteAllLedgers(LogStorage ledger) {
+        synchronized (ledger) {
+            ledger.dataMap.entrySet().stream().forEach(entry -> deleteLedger(entry.getValue().getLedgerHandle()));
+        }
+    }
+
+    //endregion
+
+
+    public LedgerData createLedgerAt(String streamSegmentName, int offset) {
+
+        log.info("Creating ledger for {} at {}", streamSegmentName, offset);
+        try {
+            LedgerHandle ledgerHandle = bookkeeper.createLedger(config.getBkEnsembleSize(),
+                    config.getBkWriteQuorumSize(), LEDGER_DIGEST_TYPE, config.getBKPassword());
+            LedgerData lh = new LedgerData(ledgerHandle, offset, 0, this.containerEpoch.get());
+            zkClient.create().forPath(getZkPath(streamSegmentName) + "/" + offset, lh.serialize());
+            return lh;
+        } catch (Exception e) {
```

Comment:
Is there more than one error here? Do they need to be handled differently?

Reply:
Curator has taken care of transient ZK errors. We can not do anything but pass the errors on.